### ISSUE TITLE
Reloading compiled CSS

### ISSUE THREAD

**Issue by mattrossman** [11ty/eleventy-dev-server#19]
Related: #18

Whereas that issue was a bug report, this is more of a feature request / discussion. The current domdiff CSS reload behavior is tricky to get working when using a CSS processor. My specific use case is to use Tailwind CSS via PostCSS. 

From what I can tell, the dev server only hot reloads a stylesheet if it's **not** being ignored by Eleventy. That's reasonable; However, when compiling CSS (either inside or outside of the Eleventy pipeline), developers will want to ignore the CSS build files in version control. By listing these build files in `.gitignore`, Eleventy will ignore it too (by default) and the dev server won't reload the built CSS.

As a workaround, I can disable the default ignore behavior with `eleventyConfig.setUseGitIgnore(false)` and then re-add everything else from `.gitignore`, either by maintaininly a nearly-identical copy of `.gitignore` inside `.eleventyignore`, or perhaps with some additional engineering in the config, but that seems like an unnecessary hoop to jump through. It's also unclear if that works at all when building CSS directly into the `_site` output folder (it seems the broader `_site` entry in `.gitignore` overpowers my attempts at excluding the CSS files with the `!` syntax).

Some possible solutions I can think of:

* Instead of listening for signals from the Eleventy build process (which is affected by the watch / ignore list), just watch the output directory for changes since those are ultimately the files that will be served.
* Expose some API that allows plugins to trigger an asset reload. In my case, I would want to signal to the server that the CSS file has been rebuilt from my "eleventy.before" PostCSS handler.
* Provide an Eleventy API for opting-out a specific entry from the ignore list. Currently I can't say `eleventyConfig.ignores.remove("_site/css/")` because that specific entry was never technically added to the ignores Set. It also seems that `.addWatchTarget()` gets overpowered by the ignores.

Of those, the first option seems most straightforward (no API changes) at the potential cost of performance.

## Additional Context

[See this discussion on Discord](https://discord.com/channels/741017160297611315/872911930120110190/956245439370362943) for additional context.

Also, here is a small repo that demonstrates what I'm trying to accomplish. It runs PostCSS in an `"eleventy.before"` handler and writes the output into the `_site` folder. With domdiff enabled, these compiled styles don't reload without a full refresh.
https://github.com/mattrossman/eleventy-domdiff-postcss/

**Comment by jeromecoupe** [comment]
I can confirm that this happens to me as well

- Eleventy 2.0.0-canary.4
- Eleventy Dev server: 1.0.0-canary.7 (bundled)

Steps to reproduce

- generate a CSS file in your build folder (NPM scripts and Sass) and link it to a page / layout
- use `npx @11ty/eleventy --serve`
- when the CSS file changes without being processed by Eleventy, the dev server does not reload the page or inject changes

Adding `.addWatchTarget()` or removing `dist` from `.gitignore` does not seem to change anything for me.

In the spirit of this lightweight dev server to be used outside of 11ty project, a straightforward solution might be to add a `watchTargets` key to `setServerOptions` that can take an array of globs / files.

```js
eleventyConfig.setServerOptions({
  port: 3000,
  showAllHosts: true,
  watchTargets: ["./dist/**/*"],
});
```

**Comment by mattrossman** [comment]
> ...in the spirit of making this usable outside of an 11ty context, it would make sense to me to make watched files explicit in the setServerOptions options with a key that can take an array of globs. Watching the output folder would likely be a good default.

I like this idea. Watching output files makes the most sense to me.

I will also point out that the current behavior affects even simpler CSS pipelines, such as using CSS concatenation in templates: https://www.11ty.dev/docs/quicktips/concatenate/

I can only get CSS reloading to work for template files with a `.css` input file. For example, consider the official example for adding SASS support: https://www.11ty.dev/docs/languages/custom/#example-add-sass-support-to-eleventy

If I change the `"scss"` extension to `"css"`, then hot reloading works for SASS syntax in `.css` templates.

```js
const sass = require("sass")

eleventyConfig.addTemplateFormats("css")  // Changed from "scss"

eleventyConfig.addExtension("css", {  // Changed from "scss"
  outputFileExtension: "css",

  compile: async function (inputContent) {
    let result = sass.compileString(inputContent)

    return async (data) => {
      return result.css
    }
  },
})
```

**Comment by zachleat** [comment]
Related #2 which would need to pull in chokidar (currently on core) for standalone use here (when eleventy core is not in play)

However, for this one specifically it seems like we can fix whatever bugs are happening with the standard `addWatchTarget` API

**Comment by zachleat** [comment]
> Instead of listening for signals from the Eleventy build process (which is affected by the watch / ignore list), just watch the output directory for changes since those are ultimately the files that will be served.

As #2 has been coded (pending a 1.0.0-canary.10 release) the above will be possible via `npx @11ty/eleventy-dev-server --input=_site` (or whatever your output directory is

**Comment by zachleat** [comment]
I do still want to investigate the `addWatchTarget` bugs seemingly happening here, but just an FYI above

**Comment by jeromecoupe** [comment]
Chiming in to say my CSS still does not reload with `v2.0.0-canary.14` even with the following config:

```js
module.exports = function (eleventyConfig) {
  // server options
  eleventyConfig.setServerOptions({
    port: 3000,
    input: "./dist",
  });

  // override base config
  return {
    dir: {
      input: "src",
      output: "./dist",
      markdownTemplateEngine: "njk",
    },
  };
};
```

Both JS and CSS are gerated via NPM scripts. What's strange is that 11ty seems to be picking up JS files changes. Can provide a reduced test case in a repo if that helps.

**Comment by jeromecoupe** [comment]
Related: #42

**Comment by zachleat** [comment]
A ha, after revisiting this one I see what’s going on, maybe in two different ways:

Eleventy uses input files to trigger reloads. For future me, consider a hypothetical Eleventy project with `index.html` and `_site/test.css` (to emulate a generated CSS file):.

Saving `index.html` triggers a reload but the domdiff does not apply any changes to the `_site/test.css` file. (`domdiff: false` works around this issue)

Saving `_site/test.css` does nothing because neither Eleventy nor the Dev Server are watching anything in the output folder. Worse, `eleventyConfig.addWatchTarget("_site/*.css");` does not fix this as the watcher ignores everything in the output folder `_site/**` by default to avoid infinite build loops.

I do like the `watchFiles` addition suggested! `watchFiles: ["./dist/**/*", "./other/dir/**/*"]` I think that could be a viable way to watch things in the output folder.

**ALL OF THAT SAID**, this works now and solves both issues above:

concurrently "npx @11ty/eleventy --watch" "npx @11ty/eleventy-dev-server --input=_site"

**Comment by jeromecoupe** [comment]
@zachleat Didn't think about decoupling Eleventy and Eleventy dev server. I can confirm that this solution works.

- Pro: keep my assets pipeline separated from Eleventy (instead of a more integrated approach that uses Eleventy to generate assets using NPM packages and JS).
- Pro: keep Eleventy builds to a minimum
- Pro: solution works when changing an asset manually as well, which means that will allow me to use ELeventy dev server in non Eleventy contexts (which was a goal of the package I think).

The main thing I miss here would be to be able to watch multiple folders ... for example when using a classic CMS, I could watch generated assets folders **and** my template folder for example. Add a --proxy option to be able to proxy any local domain and Eleventy Dev Server would indeed be a very nice arrow in the proverbial quiver and a worthy alternative to Browsersync.

**Comment by zachleat** [comment]
The `watch` option will be included with Dev Server 1.0.

You can pass in an array of globs, matching chokidar syntax.

Note that there is one outstanding “bug” here that I’d like to smooth over after things calm down a bit over here: https://github.com/11ty/eleventy-dev-server/issues/44

**Comment by zachleat** [comment]
Fixed with 0e3d47f65fe444edd540622cc42dd3fc2d47b453
